js/dataframe/cell.tsx

```diff
@@ -0,0 +1,279 @@
+  EditFailure: "EditFailure",
+  Editing: "Editing",
+  Ready: "Ready",
+} as const;
+const CellStateClassMap = {
+  EditSaving: "edit-waiting",
+  EditSuccess: "edit-success",
+  EditFailure: "edit-failure",
+  Editing: "editing",
+  Ready: "ready",
+} as const;
+export type CellState = keyof typeof CellStateMap;
+
+interface TableBodyCellProps {
+  id: string | null;
+  cell: Cell<unknown[], unknown>;
+  columns: readonly string[];
+  canEdit: boolean;
+  editRowIndex: number | null;
+  editColumnIndex: number | null;
+  setEditRowIndex: (index: number | null) => void;
+  setEditColumnIndex: (index: number | null) => void;
+  cellEditMap: Map<string, { value: string; state: CellState }>;
+  setData: (fn: (draft: unknown[][]) => void) => void;
+  setCellEditMap: (
+    fn: (draft: Map<string, { value: string; state: CellState }>) => void
+  ) => void;
+  maxRowSize: number;
+}
+
+export const TableBodyCell: FC<TableBodyCellProps> = ({
+  id,
+  cell,
+  columns,
+  canEdit,
+  editRowIndex,
+  editColumnIndex,
+  setEditRowIndex,
+  setEditColumnIndex,
+  cellEditMap,
+  setData,
+  setCellEditMap,
+  maxRowSize,
+}) => {
+  const rowIndex = cell.row.index;
+  const columnIndex = cell.column.columnDef.meta!.colIndex;
+
+  const initialValue = cell.getValue();
+  // We need to keep and update the state of the cell normally
+  const [value, setValue] = useImmer(initialValue);
+  const inputRef = useRef<HTMLInputElement | null>(null);
+
+  const [cellState, setCellState] = useState<CellState>(CellStateMap.Ready);
+
+  useEffect(() => {
+    const cellIsEditable =
+      editRowIndex === rowIndex && editColumnIndex === columnIndex;
+    if (cellIsEditable) {
+      setCellState(CellStateMap.Editing);
+    }
+  }, [editRowIndex, editColumnIndex, rowIndex, columnIndex]);
+
+  const hasUpdated = cellEditMap.has(`[${rowIndex}, ${columnIndex}]`);
+
+  const tableCellClass = CellStateClassMap[cellState];
+
+  const [errorTitle, setErrorTitle] = useState<string | undefined>(undefined);
+
+  const resetEditInfo = React.useCallback(() => {
+    setEditRowIndex(null);
+    setEditColumnIndex(null);
+  }, [setEditRowIndex, setEditColumnIndex]);
+
+  const onEsc = (e: ReactKeyboardEvent<HTMLInputElement>) => {
+    if (e.key !== "Escape") return;
+    // Prevent default behavior
+    e.preventDefault();
+
+    // inputRef.current.blur();
+    resetEditInfo();
+    // TODO-barret-future; Set focus to table? (state: Editing was aborted)
+  };
+  const onTab = (e: ReactKeyboardEvent<HTMLInputElement>) => {
+    if (e.key !== "Tab") return;
+    // Prevent default behavior
+    e.preventDefault();
+
+    const hasShift = e.shiftKey;
+
+    const newColumnIndex = editColumnIndex! + (hasShift ? -1 : 1);
+    if (newColumnIndex < 0 || newColumnIndex >= columns.length) {
+      // If the new column index is out of bounds, quit
+      return;
+    }
+
+    setEditColumnIndex(newColumnIndex);
+  };
+  const onEnter = (e: ReactKeyboardEvent<HTMLInputElement>) => {
```

Comment:
Make sure this works when scrolled out of view.

Reply:
TODO future: Make Cmd-Enter add a newline in a cell.

Reply:
I can not get this to work as there is no keyboard event to record as there is not textarea DOM element to listen to.

We may have to set focus to a "hidden" input... feels wonky. But so does losing focus when it's offscreen.

Reply:
Added note